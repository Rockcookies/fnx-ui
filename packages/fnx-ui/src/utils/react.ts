import { Children, ReactElement } from 'react';
import { isFragment } from 'react-is';

export const toElementArray = (
	children: React.ReactNode,
	option?: { keepEmpty?: boolean },
): ReactElement[] => {
	const keepEmpty = option && option.keepEmpty;
	let ret: React.ReactElement[] = [];

	Children.forEach(children, (child: any) => {
		if ((child === undefined || child === null) && !keepEmpty) {
			return;
		}

		if (Array.isArray(child)) {
			ret = ret.concat(toElementArray(child));
		} else if (isFragment(child) && child.props) {
			ret = ret.concat(toElementArray(child.props.children, option));
		} else {
			ret.push(child);
		}
	});

	return ret;
};
