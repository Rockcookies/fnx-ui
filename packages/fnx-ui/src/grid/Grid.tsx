import React, { cloneElement, forwardRef, isValidElement } from 'react';
import configComponentProps from '../hooks/config-component-props';
import { BORDER_TOP } from '../utils/constants';
import { addUnit } from '../utils/format';
import { classnames, createBEM } from '../utils/namespace';
import { toElementArray } from '../utils/react';
import { GridContext } from './context';
import { GridProps } from './interface';

const NS = 'fnx-grid';
const bem = createBEM(NS);

const useProps = configComponentProps<
	Required<
		Pick<
			GridProps,
			'columnSize' | 'clickable' | 'gutter' | 'border' | 'square'
		>
	>
>({
	columnSize: 4,
	clickable: false,
	gutter: 0,
	border: true,
	square: false,
});

const Grid = forwardRef<HTMLDivElement, GridProps>((_props, ref) => {
	const [
		{ columnSize, clickable, gutter, border, square },
		{ className, children: _children, style, ...restProps },
	] = useProps(_props);

	const children = toElementArray(_children);

	return (
		<div
			className={classnames(
				bem(),
				border && gutter <= 0 ? BORDER_TOP : undefined,
				className,
			)}
			{...restProps}
			ref={ref}
			style={{ paddingLeft: addUnit(gutter), ...style }}
		>
			<GridContext.Provider
				value={{
					columnSize,
					clickable,
					gutter,
					border,
					square,
				}}
			>
				{children.map((child, idx) => {
					if (isValidElement<any>(child)) {
						return cloneElement(child, {
							key: child.key != null ? child.key : idx,
							...child.props,
							'data-fnx-grid-item-index': idx,
						});
					} else {
						return child;
					}
				})}
			</GridContext.Provider>
		</div>
	);
});

Grid.displayName = 'Grid';

export default Grid;
