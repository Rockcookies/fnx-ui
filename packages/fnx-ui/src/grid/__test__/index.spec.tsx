import React from 'react';
import { render } from '@testing-library/react';
import Grid from '../index';

describe('<Grid/>', () => {
	const children = new Array(5)
		.fill(0)
		.map((_, idx) => <Grid.Item key={idx}>children</Grid.Item>);

	it('should render children correctly', () => {
		const { container } = render(<Grid>{children}</Grid>);

		expect(container).toMatchSnapshot();
	});

	it('should change grid gutter when using gutter prop', () => {
		const { container } = render(<Grid gutter={10}>{children}</Grid>);

		expect(container).toMatchSnapshot();
	});

	it('should change grid border when using border prop', () => {
		const { container } = render(<Grid border={false}>{children}</Grid>);

		expect(container).toMatchSnapshot();
	});

	it('should change grid square when using square prop', () => {
		const { container } = render(<Grid square={false}>{children}</Grid>);

		expect(container).toMatchSnapshot();
	});
});
