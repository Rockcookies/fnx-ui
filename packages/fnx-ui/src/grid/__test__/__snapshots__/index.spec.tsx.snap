// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Grid/> should change grid border when using border prop 1`] = `
<div>
  <div
    class="fnx-grid"
    style="padding-left: 0px;"
  >
    <div
      class="fnx-grid-item"
      style="flex-basis: 25%; padding-right: 0px;"
    >
      <div
        class="fnx-grid-item__content"
        data-fnx-grid-item-index="0"
        style="right: 0px;"
      >
        children
      </div>
    </div>
    <div
      class="fnx-grid-item"
      style="flex-basis: 25%; padding-right: 0px;"
    >
      <div
        class="fnx-grid-item__content"
        data-fnx-grid-item-index="1"
        style="right: 0px;"
      >
        children
      </div>
    </div>
    <div
      class="fnx-grid-item"
      style="flex-basis: 25%; padding-right: 0px;"
    >
      <div
        class="fnx-grid-item__content"
        data-fnx-grid-item-index="2"
        style="right: 0px;"
      >
        children
      </div>
    </div>
    <div
      class="fnx-grid-item"
      style="flex-basis: 25%; padding-right: 0px;"
    >
      <div
        class="fnx-grid-item__content"
        data-fnx-grid-item-index="3"
        style="right: 0px;"
      >
        children
      </div>
    </div>
    <div
      class="fnx-grid-item"
      style="flex-basis: 25%; padding-right: 0px; margin-top: 0px;"
    >
      <div
        class="fnx-grid-item__content"
        data-fnx-grid-item-index="4"
        style="right: 0px;"
      >
        children
      </div>
    </div>
  </div>
</div>
`;

exports[`<Grid/> should change grid gutter when using gutter prop 1`] = `
<div>
  <div
    class="fnx-grid"
    style="padding-left: 10px;"
  >
    <div
      class="fnx-grid-item"
      style="flex-basis: 25%; padding-right: 10px;"
    >
      <div
        class="fnx-grid-item__content fnx-grid-item__content--surround fnx-hairline"
        data-fnx-grid-item-index="0"
        style="right: 0px;"
      >
        children
      </div>
    </div>
    <div
      class="fnx-grid-item"
      style="flex-basis: 25%; padding-right: 10px;"
    >
      <div
        class="fnx-grid-item__content fnx-grid-item__content--surround fnx-hairline"
        data-fnx-grid-item-index="1"
        style="right: 0px;"
      >
        children
      </div>
    </div>
    <div
      class="fnx-grid-item"
      style="flex-basis: 25%; padding-right: 10px;"
    >
      <div
        class="fnx-grid-item__content fnx-grid-item__content--surround fnx-hairline"
        data-fnx-grid-item-index="2"
        style="right: 0px;"
      >
        children
      </div>
    </div>
    <div
      class="fnx-grid-item"
      style="flex-basis: 25%; padding-right: 10px;"
    >
      <div
        class="fnx-grid-item__content fnx-grid-item__content--surround fnx-hairline"
        data-fnx-grid-item-index="3"
        style="right: 0px;"
      >
        children
      </div>
    </div>
    <div
      class="fnx-grid-item"
      style="flex-basis: 25%; padding-right: 10px; margin-top: 10px;"
    >
      <div
        class="fnx-grid-item__content fnx-grid-item__content--surround fnx-hairline"
        data-fnx-grid-item-index="4"
        style="right: 0px;"
      >
        children
      </div>
    </div>
  </div>
</div>
`;

exports[`<Grid/> should change grid square when using square prop 1`] = `
<div>
  <div
    class="fnx-grid fnx-hairline--top"
    style="padding-left: 0px;"
  >
    <div
      class="fnx-grid-item"
      style="flex-basis: 25%; padding-right: 0px;"
    >
      <div
        class="fnx-grid-item__content fnx-hairline"
        data-fnx-grid-item-index="0"
        style="right: 0px;"
      >
        children
      </div>
    </div>
    <div
      class="fnx-grid-item"
      style="flex-basis: 25%; padding-right: 0px;"
    >
      <div
        class="fnx-grid-item__content fnx-hairline"
        data-fnx-grid-item-index="1"
        style="right: 0px;"
      >
        children
      </div>
    </div>
    <div
      class="fnx-grid-item"
      style="flex-basis: 25%; padding-right: 0px;"
    >
      <div
        class="fnx-grid-item__content fnx-hairline"
        data-fnx-grid-item-index="2"
        style="right: 0px;"
      >
        children
      </div>
    </div>
    <div
      class="fnx-grid-item"
      style="flex-basis: 25%; padding-right: 0px;"
    >
      <div
        class="fnx-grid-item__content fnx-hairline"
        data-fnx-grid-item-index="3"
        style="right: 0px;"
      >
        children
      </div>
    </div>
    <div
      class="fnx-grid-item"
      style="flex-basis: 25%; padding-right: 0px; margin-top: 0px;"
    >
      <div
        class="fnx-grid-item__content fnx-hairline"
        data-fnx-grid-item-index="4"
        style="right: 0px;"
      >
        children
      </div>
    </div>
  </div>
</div>
`;

exports[`<Grid/> should render children correctly 1`] = `
<div>
  <div
    class="fnx-grid fnx-hairline--top"
    style="padding-left: 0px;"
  >
    <div
      class="fnx-grid-item"
      style="flex-basis: 25%; padding-right: 0px;"
    >
      <div
        class="fnx-grid-item__content fnx-hairline"
        data-fnx-grid-item-index="0"
        style="right: 0px;"
      >
        children
      </div>
    </div>
    <div
      class="fnx-grid-item"
      style="flex-basis: 25%; padding-right: 0px;"
    >
      <div
        class="fnx-grid-item__content fnx-hairline"
        data-fnx-grid-item-index="1"
        style="right: 0px;"
      >
        children
      </div>
    </div>
    <div
      class="fnx-grid-item"
      style="flex-basis: 25%; padding-right: 0px;"
    >
      <div
        class="fnx-grid-item__content fnx-hairline"
        data-fnx-grid-item-index="2"
        style="right: 0px;"
      >
        children
      </div>
    </div>
    <div
      class="fnx-grid-item"
      style="flex-basis: 25%; padding-right: 0px;"
    >
      <div
        class="fnx-grid-item__content fnx-hairline"
        data-fnx-grid-item-index="3"
        style="right: 0px;"
      >
        children
      </div>
    </div>
    <div
      class="fnx-grid-item"
      style="flex-basis: 25%; padding-right: 0px; margin-top: 0px;"
    >
      <div
        class="fnx-grid-item__content fnx-hairline"
        data-fnx-grid-item-index="4"
        style="right: 0px;"
      >
        children
      </div>
    </div>
  </div>
</div>
`;
